### PR TITLE
feat: implement epoch change in staking

### LINKED ISSUE #735: Implement initialization of staking and system object

### PR DESCRIPTION
**Pull request by karlwuest**
closes #735

### CODE REVIEW

File: Cargo.lock
Comment (damirka) [inline Cargo.lock:1]: Is it intentional?

Comment (karlwuest) [reply Cargo.lock:1]: No, I'll exclude it. For some reason, it was replaced on my machine by the pre-commit script.

File: contracts/walrus/sources/staking/staking_inner.move [hunk 1]
```diff
@@ -284,24 +361,64 @@ public(package) fun select_committee(self: &mut StakingInnerV1, ctx: &mut TxCont
...
+
+    // Add any leftover rewards to the rewards to distribute.
+    let leftover_value = self.leftover_rewards.value();
+    rewards.join(self.leftover_rewards.split(leftover_value));
+    let rewards_per_shard = rewards.value() / (self.n_shards as u64);
+    let (node_ids, shard_assignments) = (*self.previous_committee.inner()).into_keys_values();
+    node_ids.zip_do!(
+        shard_assignments,
+        |node_id, shards| self
+            .pools[node_id]
```
Comment (damirka) [inline contracts/walrus/sources/staking/staking_inner.move:417]: I would add a TODO here: accessing pools for the second time may be expensive (not sure about the limit).

File: contracts/walrus/sources/system/epoch_parameters.move [hunk 2]
```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+module walrus::epoch_parameters;
+
+/// The epoch parameters for the system.
+public struct EpochParams has store, copy, drop {
```
Comment (damirka) [inline contracts/walrus/sources/system/epoch_parameters.move:7]: We already have these fields in `staking_pool::PoolParams`. Perhaps, worth using this struct there.

Comment (karlwuest) [reply contracts/walrus/sources/system/epoch_parameters.move:7]: They are slightly different, in particular the capacity and their meaning is different, so I would prefer using different types. The `PoolParams` also include a commission.

File: contracts/walrus/sources/staking/staking_inner.move (hunk 1 above)
Comment (karlwuest) [reply contracts/walrus/sources/staking/staking_inner.move:417]: Will add it. Related question: Do we actually need the update on the active set that is done a bit above or could we just do these updates lazily (e.g. next time someone stakes or wants to withdraw stake)?

Comment (damirka) [reply contracts/walrus/sources/staking/staking_inner.move:417]: By triggering `advance_epoch` on each pool, we make sure their voting / node parameters are in sync, which is a requirement for voting, I assume.

Comment (karlwuest) [reply contracts/walrus/sources/staking/staking_inner.move:417]: Ok, let's discuss this tomorrow. I added the TODO.

File: contracts/walrus/sources/init.move [hunk 3]
```diff
@@ -0,0 +1,42 @@
+// Copyright (c) Mysten Labs, Inc.
```
Comment (damirka) [inline contracts/walrus/sources/init.move:1]: Love this!

**Pull request merged**